### ISSUE TITLE
Multiple messages crashes node red

### ISSUE THREAD

**Issue by fschon** [MintyTrebor/moonNode#9]
I am trying to send mujltiple gcode commands to moonode. This works for just one g-code command, but if I send multiple messages then node red crashes:

Basically I have a function that creates multiple messages as follows and sends them to a moonNode CMD node:

```
const msgs = [];
msgs.push({ payload: { "cmd": "G28 X" } });
msgs.push({ payload: { "cmd": "G91" } });
msgs.push({ payload: { "cmd": "G1 X+10 F6000" } });
msgs.push({ payload: { "cmd": "G90" } });
return [msgs]; 
```

> 6 Mar 14:05:35 - [info] Starting flows
> 6 Mar 14:05:36 - [info] Started flows
> 6 Mar 14:05:44 - [red] Uncaught Exception:
> 6 Mar 14:05:44 - [error] Error: WebSocket is not open: readyState 0 (CONNECTING)
>     at WebSocket.send (/home/pi/.node-red/node_modules/node-red-contrib-moonnode/node_modules/ws/lib/websocket.js:394:13)
>     at WebSocket.open (/home/pi/.node-red/node_modules/node-red-contrib-moonnode/MoonNode.js:718:37)
>     at WebSocket.emit (node:events:513:28)
>     at WebSocket.setSocket (/home/pi/.node-red/node_modules/node-red-contrib-moonnode/node_modules/ws/lib/websocket.js:225:10)
>     at ClientRequest.<anonymous> (/home/pi/.node-red/node_modules/node-red-contrib-moonnode/node_modules/ws/lib/websocket.js:882:15)
>     at ClientRequest.emit (node:events:513:28)
>     at Socket.socketOnData (node:_http_client:574:11)
>     at Socket.emit (node:events:513:28)
>     at addChunk (node:internal/streams/readable:315:12)
>     at readableAddChunk (node:internal/streams/readable:289:9)
> nodered.service: Main process exited, code=exited, status=1/FAILURE
> nodered.service: Failed with result 'exit-code'.
> nodered.service: Consumed 33.251s CPU time.

If I just send one command it works fine:

```
const msgs = [];
msgs.push({ payload: { "cmd": "G28 X" } });
return [msgs]; 
```

Is this the wrong way to send multiple g-code commands? I cant find any examples anywhere online.

Thx.

**Comment by MintyTrebor** [comment]
Do you have the stateless flag enabled?

**Comment by fschon** [comment]
No.

**Comment by MintyTrebor** [comment]
I am not sure if they way you are constructing your multiple messages is valid. You are constructing an array, then send that array as a single message - which will not be in the standard node-red message format - I have never seen that done in node-red and I am not sure if its supported..  but I am far from an expert!

**Comment by MintyTrebor** [comment]
Have a look here : https://nodered.org/docs/user-guide/messages
It indicates, you create a single msg.payload containing an array and use the split node to turn them into individual msg.payload, which are then sent in sequence.

**Comment by fschon** [comment]
Hi Minty,

Its documented here under multiple outputs:
https://nodered.org/docs/user-guide/writing-functions
From what I understand node red will internally turn the array into
multiple discrete messages. Or so it looks like on the debug outputs.

What is the preferred way of sending multiple messages to the CMD node? Do
I use a Split node? Or does one create a macro in klipper and just send the
one command?

Thanks for the advice,
Franz

**Comment by MintyTrebor** [comment]
Let me run some tests so I can replicate it and formulate a solution.
It can normally handle a high frequency of individual msg.payload.cmd commands , which is why the stateless flag is an option on the node - to allow for a high frequency of commands.
The easiest current solution for you would be to create the macro as you suggested, and send one command to call the macro

**Comment by fschon** [comment]
I just tested a flow with a string of gcodes delimited by \n, fed into a
Split node which splits on \n, and then a function which puts the split
string into a { "cmd" : str } object and fed this to the CMD node. Same
exception gets thrown. So it appears the problem I have is not related to
the sequence of messages but something else.

However, if I put a delay/rate limiter node which restricts the message
flow to 1 every 3 seconds just before the CMD node, then it works and no
exception is thrown!

Any ideas?
Franz

**Comment by MintyTrebor** [comment]
have you tested it with stateless enabled?

**Comment by fschon** [comment]
Here's my test flow. With the rate limiter it works. If I remove the rate
limiter node, I get the exception.

[
    {
        "id": "2a2ed5e42a8fa553",
        "type": "tab",
        "label": "Flow 2",
        "disabled": false,
        "info": "",
        "env": []
    },
    {
        "id": "69399c5ad834b4e9",
        "type": "moonNode-command",
        "z": "2a2ed5e42a8fa553",
        "name": "ExecuteGCode",
        "command": "msg.payload.cmd",
        "server": "7490fa221a151ea5",
        "stateless": false,
        "x": 1349.9999732971191,
        "y": 189,
        "wires": [
            [],
            [],
            []
        ]
    },
    {
        "id": "18b01f8b4e1fe808",
        "type": "ui_button",
        "z": "2a2ed5e42a8fa553",
        "name": "",
        "group": "d92a5c8e34b4dda6",
        "order": 0,
        "width": 0,
        "height": 0,
        "passthru": false,
        "label": "Start",
        "tooltip": "Initialise antenna and home it",
        "color": "",
        "bgcolor": "",
        "className": "",
        "icon": "",
        "payload": "{\"x\":0,\"y\":0,\"z\":0}",
        "payloadType": "json",
        "topic": "topic",
        "topicType": "msg",
        "x": 230,
        "y": 140,
        "wires": [
            [
                "edc4deb818eeb28a"
            ]
        ]
    },
    {
        "id": "5e0796367cfabb03",
        "type": "moonNode-monitor",
        "z": "2a2ed5e42a8fa553",
        "name": "moonNode Monitor",
        "server": "7490fa221a151ea5",
        "autoStart": true,
        "mObjSel": [
            "stepper_enable",
            "tmc2209 stepper_x",
            "gcode_move"
        ],
        "mObjList":
"{\"objects\":[{\"object\":\"webhooks\",\"selected\":false},{\"object\":\"configfile\",\"selected\":false},{\"object\":\"mcu\",\"selected\":false},{\"object\":\"stepper_enable\",\"selected\":false},{\"object\":\"tmc2209
stepper_x\",\"selected\":false},{\"object\":\"gcode_move\",\"selected\":false},{\"object\":\"print_stats\",\"selected\":false},{\"object\":\"virtual_sdcard\",\"selected\":false},{\"object\":\"pause_resume\",\"selected\":false},{\"object\":\"display_status\",\"selected\":false},{\"object\":\"gcode_macro
CANCEL_PRINT\",\"selected\":false},{\"object\":\"gcode_macro
PAUSE\",\"selected\":false},{\"object\":\"gcode_macro
RESUME\",\"selected\":false},{\"object\":\"gcode_macro
SET_PAUSE_NEXT_LAYER\",\"selected\":false},{\"object\":\"gcode_macro
SET_PAUSE_AT_LAYER\",\"selected\":false},{\"object\":\"gcode_macro
SET_PRINT_STATS_INFO\",\"selected\":false},{\"object\":\"gcode_macro
_TOOLHEAD_PARK_PAUSE_CANCEL\",\"selected\":false},{\"object\":\"gcode_macro
_CLIENT_EXTRUDE\",\"selected\":false},{\"object\":\"gcode_macro
_CLIENT_RETRACT\",\"selected\":false},{\"object\":\"gcode_macro
_HOME_X\",\"selected\":false},{\"object\":\"gcode_macro
_HOME_Y\",\"selected\":false},{\"object\":\"gcode_macro
RETRACT_ALL\",\"selected\":false},{\"object\":\"gcode_macro
STEPPIR_ON\",\"selected\":false},{\"object\":\"gcode_macro
DISABLE_STEPPERS\",\"selected\":false},{\"object\":\"motion_report\",\"selected\":false},{\"object\":\"query_endstops\",\"selected\":false},{\"object\":\"idle_timeout\",\"selected\":false},{\"object\":\"system_stats\",\"selected\":false},{\"object\":\"manual_probe\",\"selected\":false},{\"object\":\"toolhead\",\"selected\":false}]}",
        "x": 550,
        "y": 280,
        "wires": [
            [
                "30f22464cb363f70"
            ],
            [
                "30f22464cb363f70"
            ]
        ]
    },
    {
        "id": "30f22464cb363f70",
        "type": "debug",
        "z": "2a2ed5e42a8fa553",
        "name": "MoonRaker monitor",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "payload",
        "targetType": "msg",
        "statusVal": "",
        "statusType": "auto",
        "x": 870,
        "y": 280,
        "wires": []
    },
    {
        "id": "6b636399ff76d1f9",
        "type": "split",
        "z": "2a2ed5e42a8fa553",
        "name": "",
        "splt": "\\n",
        "spltType": "str",
        "arraySplt": 1,
        "arraySpltType": "len",
        "stream": false,
        "addname": "",
        "x": 690,
        "y": 140,
        "wires": [
            [
                "e982305cadd695fc"
            ]
        ]
    },
    {
        "id": "edc4deb818eeb28a",
        "type": "function",
        "z": "2a2ed5e42a8fa553",
        "name": "GCode sequence",
        "func": "let str = ''.concat(\n    \"G28 X\\n\", \n
 \"G91\\n\",\n    \"G1 X+10 F6000\\n\",\n    \"G90\\n\");\n\nconst newMsg =
{ payload: str };\nreturn newMsg; ",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 450,
        "y": 140,
        "wires": [
            [
                "6b636399ff76d1f9"
            ]
        ]
    },
    {
        "id": "c4bc9dd1421af678",
        "type": "debug",
        "z": "2a2ed5e42a8fa553",
        "name": "debug 4",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "statusVal": "",
        "statusType": "auto",
        "x": 1320,
        "y": 100,
        "wires": []
    },
    {
        "id": "e982305cadd695fc",
        "type": "function",
        "z": "2a2ed5e42a8fa553",
        "name": "function 2",
        "func": "const newMsg = { payload: { \"cmd\": msg.payload}
};\nreturn newMsg;",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [],
        "x": 880,
        "y": 140,
        "wires": [
            [
                "98443bd1165474ea"
            ]
        ]
    },
    {
        "id": "98443bd1165474ea",
        "type": "delay",
        "z": "2a2ed5e42a8fa553",
        "name": "",
        "pauseType": "rate",
        "timeout": "5",
        "timeoutUnits": "seconds",
        "rate": "1",
        "nbRateUnits": "3",
        "rateUnits": "second",
        "randomFirst": "1",
        "randomLast": "5",
        "randomUnits": "seconds",
        "drop": false,
        "allowrate": false,
        "outputs": 1,
        "x": 1100,
        "y": 140,
        "wires": [
            [
                "c4bc9dd1421af678",
                "69399c5ad834b4e9"
            ]
        ]
    },
    {
        "id": "7490fa221a151ea5",
        "type": "moonNode-connector",
        "server": "localhost:7125"
    },
    {
        "id": "d92a5c8e34b4dda6",
        "type": "ui_group",
        "name": "Steppir control",
        "tab": "b65a296bf14e5099",
        "order": 1,
        "disp": true,
        "width": "6",
        "collapse": false,
        "className": ""
    },
    {
        "id": "b65a296bf14e5099",
        "type": "ui_tab",
        "name": "Steppir control",
        "icon": "dashboard",
        "disabled": false,
        "hidden": false
    }
]

**Comment by fschon** [comment]
I just tested with Stateless checked and it also fails.

**Comment by MintyTrebor** [comment]
OK. I'll look into it.

**Comment by MintyTrebor** [comment]
It's because the array sends the messages faster than the websocket can open for the first message in the array, so the second message errors, as the first message is still processing. I have previously tested frequency of msg's up to 20ms, so this approach must be faster than that, and faster than the web socket can open. 
I will need to add a check for this in the code.

**Comment by fschon** [comment]
Ok thx. Maybe adding a simple queue would help, or simply block the thread
and wait for the socket to be available?

How are you usually sending messages to the CMD node? The same way in the
test flow I posted? In my test flow I am now using the split node as you
suggest and not the message array and still get the exception.

I will try and set the rate limiter to a smaller value and see if that
helps, but I think the minimum is 1s which is quite a big delay.

**Comment by MintyTrebor** [comment]
TLDR ;-) 
This is going to take me a few days to resolve to my satisfaction, for now I suggest you create a macro and just send a single command to call the macro.

---

So....
There is an incongruity with my original design and some changes to node-red that I was not fully aware of the impact of. 
Node-red now handles msgs asynchronously, and the same is also true for the MoonRaker web-socket (which to be fair it always has been) - so its not really practical to have a synchronous send/receive function within an asynchronous node with a asynchronous MoonRaker web-socket.

I have made some changes to resolve the problem for your use case - but it comes with a  big caveat - you must have stateless enabled.

With my new code if you leave stateless disabled and send a array of messages to the CMD node, it will execute them in an asynchronous order, so there is no guarantee the original sequence of commands will be respected - and indeed in my tests with stateless disabled this is always the case. Enabling stateless appears to solve this problem, but even then I cannot **_guarantee_** synchronicity.

I'm not really happy with this approach, so I am thinking about switching the CMD node over to the HTTP rest API of MoonRaker, which would enforce a synchronous send/receive and avoid this problem entirely, and allow the original design/intent of the command node to be retained.

**Comment by MintyTrebor** [comment]
OK - Good News:
I have transferred over to using HTTP requests and have resolved the asynchronous message processing problem, and fixed the original issue you raised (code is a lot simpler too - which is great from my perspective).

However - I seem to have identified a MoonRaker issue (at least on my printer), where it will sometimes process gcode commands in a different sequence to how they arrived. I have proven this by:
Created a flow with a array of M118 commands : M118 'HI', M118 'HI2', M118 'HI3' etc and send to the CMD node
Here is the output from the node-red server on wireshark:
![image](https://user-images.githubusercontent.com/26109080/223700773-6d096215-06eb-495f-8a4d-d50e19645264.png)

You can see that the commands are being sent to moonraker in the correct order, but this is the output sequence the klipper console shows :
![image](https://user-images.githubusercontent.com/26109080/223701111-e52af916-999f-4bf4-98bc-fae39e282a92.png)

It does not happen all the time, but still could cause problems - so I am going to look into rate limiting.

**Comment by fschon** [comment]
Great, thanks for your efforts Minty.

Can you send me the flow with your M118 commands so are texting the same
thing pls.

**Comment by MintyTrebor** [comment]
I have just published the updated version 0.1.5. 
As I have ended up adding a queue, I have also added the ability to cancel the current queue. Please see the updated help text in node-red for more info.

To Test I am using a function node with the following :
```
const msgs = [];
msgs.push({ payload: { "cmd": "M118 'HI'" } });
msgs.push({ payload: { "cmd": "M118 'HI2'" } });
msgs.push({ payload: { "cmd": "G28" } });
msgs.push({ payload: { "cmd": "M118 'HI3'" } });
msgs.push({ payload: { "cmd": "M118 'HI4'" } });
return [msgs];
```
With the changes I have implemented, everything is now executed in sequence, and the next command in the queue does not get sent until MoonRaker has responded to the current command.

It would appear that if you send commands to moonraker without waiting for a response between each command it may process them asynchronously, which does not seem right to me....

**Comment by fschon** [comment]
Hi Minty,

Thanks for this update. I have updated to 0.1.5 and also had to update
nodejs to version 18.

However now I don't get any messages getting through to moonraker. None of
my commands have any effect. I also added a moonNode Monitor node and
configured that but when I click Refresh on the object list I get a
communication error

[image: image.png]

Is there something new on the configuration that I am missing?

Thx,
Franz

**Comment by fschon** [comment]
Hi Minty,

Thanks for this update. I have updated to 0.1.5 and also had to update
nodejs to version 18.

However now I don't get any messages getting through to moonraker. None of
my commands have any effect. I also added a moonNode Monitor node and
configured that but when I click Refresh on the object list I get a
communication error

![image](https://user-images.githubusercontent.com/18084317/224479609-5eec01da-8203-4e7c-b736-54beb2f7a24d.png)

Is there something new on the configuration that I am missing?

Thx,
Franz

**Comment by MintyTrebor** [comment]
can you try with the IP address of your pi instead of localhost... 
You will need to deploy and reload node-red before refreshing the objects (this is a node-red bug if you change connection params)

**Comment by MintyTrebor** [comment]
I have literally 5 mins ago, just updated my printer to the latest version of klipper & Moonraker, and updated my production node-red to moonnode 0.1.5 and it is working

**Comment by MintyTrebor** [comment]
if you need some real time support I am on discord (Minty Trebor#8425)

**Comment by fschon** [comment]
Setting the actual IP address did the trick thx. Don't know why localhost doesn't work anymore, but possibly had to do with the nodejs update or other updates that have been performed? But importantly my steppers are now working perfectly :-).

**Comment by MintyTrebor** [comment]
Fantastic.
